llvm/lib/Target/AIE/AIEMachineScheduler.cpp

```diff
@@ -479,27 +480,78 @@ SUnit *AIEPostRASchedStrategy::pickNodeAndCycle(
 }
 
 int AIEPostRASchedStrategy::getMaxDeltaCycles(const SchedBoundary &Zone) const {
-  assert(!Zone.isTop());
-  if (Zone.getCurrCycle() >= RegionBottomUpCycles - 1)
+  // Top-down scheduling does not support DeltaCycles
+  if (Zone.isTop() || Zone.getCurrCycle() >= RegionBottomUpCycles - 1)
     return 0;
   return std::min({int(RegionBottomUpCycles - 1 - Zone.getCurrCycle()),
                    int(getAIEHazardRecognizer(Zone)->getMaxLookAhead()),
                    BottomUpDelta.getValue()});
 }
 
+unsigned getNumEmittedInstrs(ScheduleDAGMI *DAG, bool IsTop) {
+  if (IsTop)
+    return DAG->top().isValid() ? std::distance(DAG->begin(), DAG->top()) : 0;
+  return DAG->bottom().isValid() ? std::distance(DAG->bottom(), DAG->end()) : 0;
+}
+
+SUnit *AIEPostRASchedStrategy::getNextUnscheduledFixedInstr(
+    const SchedBoundary &Zone) const {
+  if (Zone.isTop())
+    return nullptr;
+  const Region &Reg = InterBlock.getBlockState(CurMBB).getCurrentRegion();
+  unsigned NumEmitted = getNumEmittedInstrs(DAG, /*IsTop=*/false);
+  if (NumEmitted < Reg.getNumBotFixedInstructions()) {
+    MachineInstr &NextMI =
+        *std::prev(DAG->bottom().isValid() ? DAG->bottom() : DAG->end());
+    SUnit *NextSU = DAG->getSUnit(&NextMI);
+    assert(NextSU);
+    assert(NextSU->BotReadyCycle == NextSU->getHeight() &&
+           "Fixed instruction won't be placed at the correct cycle");
+    assert(Zone.getCurrCycle() <= NextSU->BotReadyCycle);
+    return NextSU;
+  }
+  return nullptr;
+}
+
+bool AIEPostRASchedStrategy::isFixedSU(const SUnit &SU, bool IsTop) const {
+  if (IsTop) {
+    return FirstTopFixedSU && SU.NodeNum >= *FirstTopFixedSU &&
+           SU.NodeNum < FirstBotFixedSU.value_or(DAG->SUnits.size());
+  }
+  return FirstBotFixedSU && SU.NodeNum >= *FirstBotFixedSU &&
+         SU.NodeNum <= LastBotFixedSU.value();
+}
+
+bool AIEPostRASchedStrategy::isFreeSU(const SUnit &SU) const {
+  const unsigned NumUpperBound = DAG->SUnits.size();
+  return SU.NodeNum < FirstTopFixedSU.value_or(NumUpperBound) &&
+         SU.NodeNum < FirstBotFixedSU.value_or(NumUpperBound);
```

Comment:
Top and Bottom look too symmetric here. Is that intentional?

Reply:
Yes, all "fixed" SUnits are created after "free" SUnits, regardless of Top/Bot. See the header:
```
  /// SU numbers for fixed instructions.
  /// "top" fixed SUnits belong in [FirstTopFixedSU,FirstBotFixedSU)
  /// "bot" fixed SUnits belong in [FirstBotFixedSU,LastBotFixedSU]
  std::optional<unsigned> FirstTopFixedSU;
  std::optional<unsigned> FirstBotFixedSU;
  std::optional<unsigned> LastBotFixedSU;
```